### DISCUSSION TITLE
Address Book enhancements

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4010] [Development]
**To Do**:

- [x] Move the Address config into libaddress
- [x] Add patterns for the Address fields
- [x] Upgrade search to allow using patterns
- [x] Add a `<limit>` function
- [ ] Move the function data to become Window data
- [x] Eliminate config global variables

**Advanced**:

- Add functions for parity with other dialogs:
  `<tag-pattern>`, `<delete-pattern>`, `<undelete-pattern>`
- Use some formatted comments, e.g. `# label=red,blue` for labelling
  These would allow users to group their aliases
- Add a sidebar of labels, including "All", "Self" (from alternates), plus each user label

**Comment by flatcap** [comment]
## Address Book - Add Patterns

Currently, the search works my matching a regex against the fulltext displayed
on the screen.  Adding the following patterns would allow more accurate
searching.

- `~f EXPR` - From (alias)
- `~t EXPR` - To (address)
- `~b EXPR` - Body (comment)
- `~Y EXPR` - Tag (label) -- not yet implemented

If the search didn't contain a `~`, then we'd use the old search method.

## Add a Limit Command

In the index, you can limit the view by using a pattern.

The Address Book is started via auto-completion on an Address.
If you type `joh<tab>` then only matching Addresses are shown.
Once in this limited view, you can't get back to the full list.

Rather than filtering the list, I suggest starting with a limit set.
This would allow the user to see the full list, by `<limit>.<enter>`

## Add Pattern Functions

Using the patterns, we can now add functions:

- `<tag-pattern>`
- `<delete-pattern>`
- `<undelete-pattern>`

**Comment by abc123098123** [comment]
Hey @flatcap!
I would like to ask you how you would approach the `Address Book - Add Patterns` feature.

I have been trying to understand how the Patterns in the index window work. And I think I did figure it out.
My problem at the moment is how to actually fit the Address Book into this.

I know that when we search in the index window, `mutt_search_command` is executed.
Inside it, if there is no cache, `mutt_pattern_comp` will be executed and it will define the op and tag of the search input (e.g. if I use `~t Romeu` as the input, it will search for the next e-mail that has `Romeu` in the `To:` field.).
Finally, `mutt_pattern_exec` will be used to check if it matches or not.

Now, my question is more about `mutt_pattern_exec`: it receives an Email struct and, if using `~t`, it will use `e->env->to` to check if it matches. How should I integrate a new struct in here, that would check the Alias Address?

Thank you very much!

**Comment by flatcap** [comment]
Hmm... It's a bit more complicated than I thought.
I'll have a closer look in the morning, but...

I think we need our own custom version of `mutt_pattern_exec()` that takes an `AliasView` instead of an `Email`.
Our version will be **much** simpler.  Plus, we'll need our own `perform_and()` and `perform_or()` glue functions.

I think we can probably reuse much of the same preparation, creating the Pattern, etc.

**Comment by abc123098123** [comment]
I guess it is feasible :-)

**Comment by flatcap** [comment]
Here's another brain dump :-)

We'll probably need:
`menu->custom_search = true;`

Starting in `mutt_index_menu()`

- `case OP_SEARCH_REVERSE:`
- `case OP_SEARCH_NEXT:`
- `case OP_SEARCH_OPPOSITE:`
- `case OP_SEARCH:`

calls `mutt_search_command()`

call `mutt_get_field()` - get a string to find

calls `mutt_check_simple()` - check for `~X` patterns
    if there aren't any `~` patterns, use the old search behaviour
    (regex against the full on-screen text)

Index stores
```c
struct PatternList *SearchPattern = NULL; ///< current search pattern
char LastSearch[256] = { 0 };             ///< last pattern searched for
char LastSearchExpn[1024] = { 0 }; ///< expanded version of LastSearch
```

`SearchPattern = mutt_pattern_comp()`

Then loop through `Alias`es,
- Start at current item `Menu.current`
- If end is reached, wrap around (`C_WrapSearch`)
- If original item is reached, stop

Search is performed by `mutt_pattern_exec()` which compares a regex against an `Email`.

We need a simple `alias_pattern_exec()` which takes an `AliasView`.

- `~f EXPR`, `MUTT_PAT_FROM` - From (alias)
- `~t EXPR`, `MUTT_PAT_TO`   - To (address)
- `~b EXPR`, `MUTT_PAT_BODY` - Body (comment)

Plus `perform_and()` and `perform_or()` glue functions.

**Comment by abc123098123** [comment]
@flatcap 

Hi!

I will be talking specifically about `Address Book - Add Patterns`.

Well, I still didn't commit my branch... but good news:

- `~f EXPR` is working;
- `~t EXPR` is working;
- `~b EXPR` is working;

But a couple things though:

1. You mentioned Tag (`~Y EXPR`) and you also mentioned it is not implemented. My suggestion is that I do this issue without the Tag, and the Tag implementation comes in a later PR. I'll do whatever you think it is best!
2. I did not have to create new versions of `perform_and()` or `perform_or()`. They are related to `MUTT_PAT_AND` and `MUTT_PAT_OR`. I don't think they have anything to do with this issue, am I correct?
3. To make this work, I changed the `menu.c` file, in the big switch-case, inside the `case` related to the search. Then I call a specific version of my function, which was made specifically for `Menu` instead of `Mailbox`. I think this is also correct.

Thanks!

**Comment by flatcap** [comment]
> 1. You mentioned Tag (`~Y EXPR`) and you also mentioned it is not implemented

Yes, not yet.  **If** we go ahead with the Advanced ideas, we'll need it.

> 2. I did not have to create new versions of `perform_and()` or `perform_or()`.
> They are related to `MUTT_PAT_AND` and `MUTT_PAT_OR`.

I think we do need these.
The `SearchPattern` is a **list** of terms.
By default, these terms are AND'd together: `~a ~b` implies `~a` **AND** `~b`.
`mutt_pattern_comp()` checks for `|`, meaning that `SearchPattern`s terms should be OR'd.
The function also deals with `()`s for precedence.

> 3. To make this work, I changed the `menu.c` file

Hmm... I don't think we want _that_.
When we `menu_new()`, if we set `menu->custom_search = true`, then `OP_SEARCH` will be returned to `dlg_select_alias()`.
Then you can call your function.

**Comment by abc123098123** [comment]
@flatcap 

Hey!

I have been studying the `<limit>` function that is used inside index.

`mutt_pattern_func` uses Context struct extensively, creating the `v2r` mapping.
`mutt_draw_tree` uses `Context->threads`, which is a `ThreadsContext` related to the Mailbox.

Context struct is heavily used here, and it seems it is mailbox specific, according to the docstring (`struct Context - The "current" mailbox`).

Now that we are trying to introduce `<limit>` in a new area, which is *not* a Mailbox, how should we proceed?

Should there be a `AliasContext` inside `Context`, just like `ThreadsContext`? (Although `ThreadsContext` is used for the thread specific limits, it is also used in general limits).

Should there be a parallel `Context` like struct but for the Alias area?

What would you recommend?

Thanks!

**Comment by flatcap** [comment]
Hmm... I started by writing...

> Good news.  It's nowhere as complicated as that.

then got a bit worried.

I just need to check a few things...

**Comment by flatcap** [comment]
We want to control what the `Menu` displays, but the it doesn't hold the data, we do.
`Menu.data` is our `ARRAY` of `AliasView`.  We just need to make some tweaks.

First, we'll add a `bool is_visible` to `AliasView` -- limiting the view will affect this.

We'll also need to store a couple more items of data:
- The limit string that the use typed in
- The compiled `Pattern` generated from that string

I suggest shuffling things around:
- `AliasMenuData` -> `AliasViewArray`
- New `struct AliasMenuData`, containing the `ARRAY` and the new data

`When the user types in a limit we:
- Create a `Pattern`
- Match it against all the Aliases
- Set `is_visible` accordingly

Then we do a bit of creative sorting.

Change the sort functions so that Aliases that aren't visible sort to the end of the list.
Now recount the visible Aliases and update the count in the `Menu`.

When the user clears the limit, set all the Aliases to visible and resort them.
Set the count in the `Menu` to ARRAY_SIZE().

**Comment by flatcap** [comment]
# Alias Tests

These tests are for the Address Book (Alias-selection Dialog).
Most should apply to the Query Dialog too.

## Data

The [sample-data](https://github.com/neomutt/sample-data) repo has Actor data in alias and query format.
I suggest randomising the files with `shuf` so that sorting has a visible effect.

Querying can be done using a simple script:
```sh
echo "# Header"
grep -i "$1" actor-query.txt
```

## Completion

The Address Book can be reached by auto-completing an Address when composing an email.

<kbd>m</kbd> (to `<mail>`), then <kbd>joh</kbd> + <kbd>\<tab\></kbd> to complete.

### Empty Tests

Start NeoMutt with no config, then try the Address Book.

```
neomutt -n -F /dev/null
```

| Type in   | Expected Result | Notes  |
| :-------- | :-------------- | :----- |
| [nothing] | Nothing happens |        |
| jo        | Nothing happens |        |

### Data Tests

| Type in   | Expected Result                 | Notes                                  |
| :-------- | :------------------------------ | :------------------------------------- |
| [nothing] | All Aliases displayed           | No filtering done                      |
| jo        | Aliases starting 'jo' displayed | Limit applied, multiple matches        |
| ang       | Completes to 'angelina'         | Only one match, Address Book not shown |
| angelina  | One Alias shown                 | Limit applied, only one match          |
| garbage   | All Aliases displayed           | No matches                             |

## Sorting

Test each of the sort methods and each way to set the sort.

- `set alias_sort = address`
- `set alias_sort = alias`
- `set alias_sort = unsorted`
- `set alias_sort = reverse-address`
- `set alias_sort = reverse-alias`
- `set alias_sort = reverse-unsorted`

Repeat using <kbd>o</kbd> (`<sort-alias>`) and <kbd>O</kbd> (`<sort-alias-reverse>`)

## Patterns

The searching, limiting and sorting tests will follow the same pattern.

| Sample Test         | Description                           |
| :------------------ | :------------------------------------ |
| john                | Plain search multiple matches         |
| angel               | Plain search single match             |
| blu                 | Plain search no matches               |
|                     |                                       |
| ~f john             | Pattern search multiple matches       |
| ~f angel            | Pattern search single match           |
| ~f blu              | Pattern search no matches             |
|                     |                                       |
| ~f john  ~c oscar   | Pattern search (AND) multiple matches |
| ~f ant   ~c oscar   | Pattern search (AND) single match     |
| ~f blu   ~c oscar   | Pattern search (AND) no matches       |
|                     |                                       |
| ~f john \| ~c oscar | Pattern search (OR) multiple matches  |
| ~f ant  \| ~t hop   | Pattern search (OR) single match      |
| ~f blu  \| ~c nobel | Pattern search (OR) no matches        |

## Searching

Using each of the Patterns above, try searching the Aliases.

- Search forwards <kbd>/</kbd>
- Search backwards <kbd>Alt-/</kbd>
- After a successful search
  - <kbd>n</kbd> (`<search-next>`)
  - <kbd>N</kbd> (`<search-opposite>`)
- After a failed search
  - <kbd>n</kbd> (`<search-next>`)
  - <kbd>N</kbd> (`<search-opposite>`)

## Limiting

Using each of the Patterns above, try limiting the Aliases.

Check that menu navigation works as expected.

- <kbd>j</kbd> (`<next-entry>`)
- <kbd>n</kbd> (`<previous-entry>`)
- <kbd>\<PageDown\></kbd> (`<next-page>`)
- <kbd>\<PageUp\></kbd> (`<previous-page>`)
- etc

Check that sorting works correctly whilst the view is limited.
The Aliases will be reordered, but the amount won't change.

## Updates

Test changes to the Aliases.

### Add Aliases

When viewing the Address Book, add new aliases.

```
alias bobh Bob Hoskins <bh@fig.com>
```

The new aliases should:
- Appear immediately (if on screen)
- Respect the sort order

Now, limit the view and test:
- New Alias that matches the limit
- New Alias that doesn't match the limit

### Remove Aliases

When viewing the Address Book, remove aliases.

```
unalias art
```

The removed aliases should:
- Disappear immediately (if on screen)

Now, limit the view and test:
- Remove an Alias that matches the limit
- Remove an Alias that doesn't match the limit

Finally, test `unalias *` with and without a limit.

**Comment by abc123098123** [comment]
Hey!

`- Move the function data to become Window data`

Could you clarify this ?

Thanks!

**Comment by flatcap** [comment]
See: [Config Elimination](https://github.com/neomutt/neomutt/issues/2622#issuecomment-701314444)
